src/formula/variant_private.hpp

```diff
@@ -0,0 +1,466 @@
+	}
+
+	virtual std::string get_debug_string(const_formula_callable_vec& /*seen*/, bool /*verbose*/) const override
+	{
+		return to_string_impl(true);
+	}
+
+	virtual bool operator==(variant_value_base& other) const override
+	{
+		return value_ == value_ref_cast<variant_decimal>(other).value_;
+	}
+
+	virtual bool operator<=(variant_value_base& other) const override
+	{
+		return value_ <= value_ref_cast<variant_decimal>(other).value_;
+	}
+
+	virtual const VARIANT_TYPE get_type() const override
+	{
+		return VARIANT_TYPE::TYPE_DECIMAL;
+	}
+
+private:
+	std::string to_string_impl(const bool sign_value) const;
+
+	int value_;
+};
+
+
+class variant_callable : public virtual variant_value_base
+{
+public:
+	explicit variant_callable(const formula_callable* callable);
+
+	virtual bool as_bool() const override
+	{
+		return callable_ != nullptr;
+	}
+
+	virtual size_t num_elements() const override
+	{
+		return 1;
+	}
+
+	const_formula_callable_ptr get_callable() const
+	{
+		return callable_;
+	}
+
+	formula_callable_ptr get_callable_mutable()
+	{
+		return mutable_callable_;
+	}
+
+	virtual std::string string_cast() const override
+	{
+		return "(object)";
+	}
+
+	virtual std::string get_serialized_string() const override;
+
+	virtual std::string get_debug_string(const_formula_callable_vec& seen, bool verbose) const override;
+
+	virtual bool operator==(variant_value_base& other) const override;
+	virtual bool operator<=(variant_value_base& other) const override;
+
+	virtual const VARIANT_TYPE get_type() const override
+	{
+		return VARIANT_TYPE::TYPE_CALLABLE;
+	}
+
+private:
+	const_formula_callable_ptr callable_;
+	formula_callable_ptr mutable_callable_;
+};
+
+
+class variant_string : public virtual variant_value_base
+{
+public:
+	explicit variant_string(const std::string& str) : string_(str) {}
+
+	virtual bool is_empty() const override
+	{
+		return string_.empty();
+	}
+
+	virtual bool as_bool() const override
+	{
+		return !is_empty();
+	}
+
+	const std::string& get_string() const
+	{
+		return string_;
+	}
+
+	virtual std::string string_cast() const override
+	{
+		return string_;
+	}
+
+	virtual std::string get_serialized_string() const override;
+
+	virtual std::string get_debug_string(const_formula_callable_vec& /*seen*/, bool /*verbose*/) const override
+	{
+		return string_;
+	}
+
+	virtual bool operator==(variant_value_base& other) const override
+	{
+		return string_ == value_ref_cast<variant_string>(other).string_;
+	}
+
+	virtual bool operator<=(variant_value_base& other) const override
+	{
+		return string_ <= value_ref_cast<variant_string>(other).string_;
+	}
+
+	virtual const VARIANT_TYPE get_type() const override
+	{
+		return VARIANT_TYPE::TYPE_STRING;
+	}
+
+private:
+	std::string string_;
+};
+
+
+template<typename T>
+class variant_container : public virtual variant_value_base
+{
+public:
+	explicit variant_container(const T& container)
+		: container_(container)
+		, container_iter_(container_.begin())
```

Comment:
Pretty sure there's no guarantee that `begin()` will be valid after inserting some elements. (Possibly, even after removing some.) So uh... it's probably a bad idea to be actually storing it like this.